src/Json/BitzArt.Json.TypedValues/Models/TypedValue.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System.Diagnostics;
+using System.Text.Json.Serialization;
+
+namespace System.Text.Json;
+
+/// <summary>
+/// Persists actual type information for a value upon JSON serialization.
+/// </summary>
+[JsonConverter(typeof(TypedValueJsonConverter))]
+[DebuggerDisplay("{Value}")]
+public abstract class TypedValue
+{
+    private object? _value;
+
+    /// <summary>
+    /// The value that requires persisting type information.
+    /// </summary>
+    public object? Value
+    {
+        get => GetValue();
+        set => SetValue(value);
+    }
+
+    private protected virtual object? GetValue() => _value;
+    private protected virtual void SetValue(object? value) => _value = value;
+
+    internal TypedValue()
+    {
+        _value = null!;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TypedValue"/> class.
+    /// </summary>
+    /// <param name="value">The value that requires persisting type information.</param>
+    internal TypedValue(object value)
+    {
+        _value = value;
+    }
+
+    /// <inheritdoc/>
+    public override int GetHashCode()
+    {
+        if (Value is null)
+        {
+            return 0;
+        }
+
+        return Value.GetHashCode();
+    }
+
+    /// <inheritdoc/>
+    public override string ToString()
+    {
+        return Value?.ToString() ?? string.Empty;
+    }
+
+    /// <inheritdoc/>
+    public override bool Equals(object? obj)
+    {
```

Comment:
Little bit confused with this logic, because we have a mix of comparison of wrapper and wrapped value, as I understand it